starlite/dto/backends/msgspec.py

```diff
@@ -0,0 +1,76 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, NewType
+from uuid import uuid4
+
+from msgspec import Struct, defstruct, field, to_builtins
+
+from starlite.dto.types import NestedFieldDefinition
+from starlite.enums import MediaType
+from starlite.exceptions import SerializationException
+from starlite.serialization import decode_json, decode_msgpack
+from starlite.types import Empty
+
+from .abc import AbstractDTOBackend
+
+if TYPE_CHECKING:
+    from starlite.dto.types import FieldDefinition, FieldDefinitionsType
+
+__all__ = ["MsgspecDTOBackend"]
+
+
+MsgspecField = NewType("MsgspecField", type)
+
+
+class MsgspecDTOBackend(AbstractDTOBackend[Struct]):
+    def parse_raw(self, raw: bytes, media_type: MediaType | str) -> Any:
+        if media_type == MediaType.JSON:
+            transfer_data = decode_json(raw, type_=self.annotation)
+        elif media_type == MediaType.MESSAGEPACK:
+            transfer_data = decode_msgpack(raw, type_=self.annotation)
+        else:
+            raise SerializationException(f"Unsupported media type: '{media_type}'")
+        return to_builtins(transfer_data)
```

Comment:
Thanks for taking a look @jcrist.

My angle was to have this method return primitives so that the conversion from Struct/BaseModel to e.g., SQLAlchemy or dataclass or whatever could be standardized.

I also notice that using `to_builtins` will convert `UUID` and `datetime` objects back to their string form, which isn't what I want either.

Will def. be better to go straight from struct -> SQLAlchemy or whatever the model is.

Cheers:)

Reply:
I probably don't understand what this abstraction is handling here (I'm not really a starlite user, just trying to help users have success when using `msgspec`). Just a few assorted notes about things `msgspec` can do that you might find useful for this work (apologies if you're already aware of these):

- `msgspec.json.decode` can decode a JSON message directly into `dataclasses`, `attrs`, or `msgspec.Struct` types, complete with validation. If the end output type is one of those, it would be most performant to pass the type directly to `msgspec.json.decode`, rather than parsing, creating a dict, then converting.
- `msgspec.json.decode` can decode a JSON message directly into a `TypedDict` schema. If the end output type is a `dict`, then this will be slightly more performant than decoding into a structured object, then converting to a `dict`. However, the perf diff here should be pretty small, and `TypedDict` annotations are less feature rich (no default values, for example).
- If you're trying to convert a `msgspec.Struct` into a dict with _no_ recursion or conversion of the structs fields, you can do this with `msgspec.structs.asdict` ([docs](https://jcristharif.com/msgspec/api.html#msgspec.structs.asdict)).

Reply:
> just trying to help users have success when using msgspec

Really appreciated!

> I probably don't understand what this abstraction is handling here

Converting raw bytes into some domain modelling type and then back to bytes again, where the to/from bytes conversions are performed internally by Starlite.

We generate the `Struct` definitions internally given the users domain model and a configuration, e.g., in this example, `Author` is a sqlalchemy model.

```py
WriteDTO = SQLAlchemyDTO[Annotated[Author, DTOConfig(exclude={"id"})]]
```
Internally, the `SQLAlchemyDTO` type will generate a `Struct`<sup>1</sup> that includes field definitions equivalent to the columns defined on the sqlalchemy model, but excluding the `id` attribute. That struct is then used to parse/validate the raw data received by a handler, and then we create an `Author` instance populated with data from the `Struct` instance and inject that into the handler.

I wasn't familiar with `asdict()`, I think that will be useful.

1: Actually the implementation you can see here is generating pydantic models internally, but in my local branch I've switched to using structs. In starlite v1 the DTO types were subclasses of `BaseModel` themselves, whereas in 2.0 they are their own type that holds a reference to a ser/de model, instead of actually _being_ one.